### ISSUE TITLE
Can we add response predicate to the retry config as well  ?

### ISSUE THREAD

**Issue by Romeh** [resilience4j/resilience4j#259]
Hey All ,
Small question , can we add a support to have a response predicate of the retry logic beside the exceptions ? like retry if the response of a service call or a specific method output , as some cases we need to retry if we get specific response or status , not only having exceptions being thrown ,  if we need to explicitly throw an exception to retry based into the response result validation , i do not think it is a good practice to use exception to control your functional flow .

Can you please advise me on that ? I can try to do the code a change and create a pull request for that as well if you want .

Thanks a lot

**Comment by Romeh** [comment]
@storozhukBM  I agree with you that it is not a general truth , but doing retry on response predicate for me is part of business logic and not an exceptional action , as for exceptional action it means a showstopper and you need to break which can be for example if you kept retrying and failed at the end to fulfill the successful end condition . 

Some example of what i mean from similar libraries but i like resilience4j more :
https://github.com/elennick/retry4j
https://github.com/jhalterman/failsafe
https://github.com/rholder/guava-retrying --> really old

What do you think ?

**Comment by Romeh** [comment]
Dears , 

I have the added the implementation in my fork for your review as being mentioned above with the commit id , 

Basically the DSL for retry config will be extended to support response predicate for sync and Async retry if any defined :

Example: 
// Create a Retry with default configuration
`final RetryConfig retryConfig = RetryConfig<String>custom().retryOnResult(s->s.contains("NoRetry")).maxAttempts(1).build();`

I have added many unit testing cases as well in AsyncRetryTest and SupplierRetryTest classes 

Can you please review and tell me what do you think about it if there is any issues or corner cases i did not tackle properly

**Comment by RobWin** [comment]
Thanks, I have a look as soon as I have some time.
